File: src/auth/oauth-provider.ts
import { ApiError, isAbortError } from "../shared/api-error";
import { isRecord, stringField, type JsonObject } from "../shared/json";
import {
	BodySizeLimitError,
	readLimitedBody,
} from "../shared/limited-body";

const CODEX_CLIENT_ID = "app_EMoamEEZ73f0CkXaXp7hrann";
const DEVICE_USER_CODE_URL =
	"https://auth.openai.com/api/accounts/deviceauth/usercode";
const DEVICE_TOKEN_URL =
	"https://auth.openai.com/api/accounts/deviceauth/token";
export const DEVICE_VERIFICATION_URL = "https://auth.openai.com/codex/device";
const DEVICE_TOKEN_REDIRECT_URI =
	"https://auth.openai.com/deviceauth/callback";
const OAUTH_TOKEN_URL = "https://auth.openai.com/oauth/token";
const MAX_OAUTH_RESPONSE_BYTES = 64 * 1024;
const PROVIDER_REQUEST_TIMEOUT_MS = 10_000;

interface ProviderDeviceAuthorization {
	deviceAuthId: string;
	userCode: string;
	interval: unknown;
}

export type ProviderDevicePollResult =
	| { status: "pending" }
	| { status: "authorized"; tokenPayload: unknown };

class OAuthProviderFailure extends Error {
	constructor(readonly status: number | undefined) {
		super("OAuth provider request failed.");
		this.name = "OAuthProviderFailure";
	}
}

interface ProviderRequestOptions {
	body: JsonObject | URLSearchParams;
	signal?: AbortSignal;
	networkError?: () => Error;
}

interface ProviderResponse {
	response: Response;
	clientSignal?: AbortSignal;
	networkError: () => Error;
}

export async function requestDeviceAuthorization(
	signal?: AbortSignal,
): Promise<ProviderDeviceAuthorization> {
	const upstream = await providerFetch(DEVICE_USER_CODE_URL, {
		body: { client_id: CODEX_CLIENT_ID },
		...(signal ? { signal } : {}),
	});
	const payload = await requireProviderJson(upstream);
	if (!isRecord(payload)) throw invalidProviderResponse();
	const deviceAuthId = stringField(payload, "device_auth_id");
	const userCode =
		stringField(payload, "user_code") ?? stringField(payload, "usercode");
	if (!deviceAuthId || !userCode) throw invalidProviderResponse();
	return { deviceAuthId, userCode, interval: payload.interval };
}

export async function pollDeviceAuthorizationToken(
	deviceAuthId: string,
	userCode: string,
	signal?: AbortSignal,
): Promise<ProviderDevicePollResult> {
	const upstream = await providerFetch(DEVICE_TOKEN_URL, {
		body: {
			device_auth_id: deviceAuthId,
			user_code: userCode,
		},
		...(signal ? { signal } : {}),
	});

	if (upstream.response.status === 403 || upstream.response.status === 404) {
		await discardBody(upstream.response);
		return { status: "pending" };
	}
	const payload = await requireProviderJson(upstream);
	if (!isRecord(payload)) throw invalidProviderResponse();
	const authorizationCode = stringField(payload, "authorization_code");
	const codeVerifier = stringField(payload, "code_verifier");
	const codeChallenge = stringField(payload, "code_challenge");
	if (!authorizationCode || !codeVerifier || !codeChallenge) {
		throw invalidProviderResponse();
	}

	return {
		status: "authorized",
		tokenPayload: await exchangeAuthorizationCode(
			authorizationCode,
			codeVerifier,
			signal,
		),
	};
}

export async function refreshProviderToken(
	refreshToken: string,
): Promise<unknown> {
	try {
		return await exchangeRefreshToken(refreshToken);
	} catch (error) {
		if (error instanceof ApiError) throw error;
		throw providerApiError(
			error instanceof OAuthProviderFailure ? error.status : undefined,
		);
	}
}

async function exchangeAuthorizationCode(
	code: string,
	codeVerifier: string,
	signal?: AbortSignal,
): Promise<unknown> {
	const upstream = await providerFetch(OAUTH_TOKEN_URL, {
		body: new URLSearchParams({
			grant_type: "authorization_code",
			client_id: CODEX_CLIENT_ID,
			code,
			redirect_uri: DEVICE_TOKEN_REDIRECT_URI,
			code_verifier: codeVerifier,
		}),
		...(signal ? { signal } : {}),
	});
	return requireProviderJson(upstream);
}

async function exchangeRefreshToken(refreshToken: string): Promise<unknown> {
	for (let attempt = 0; attempt < 3; attempt++) {
		try {
			const upstream = await providerFetch(OAUTH_TOKEN_URL, {
				body: new URLSearchParams({
					client_id: CODEX_CLIENT_ID,
					grant_type: "refresh_token",
					refresh_token: refreshToken,
					scope: "openid profile email",
				}),
				networkError: () => new OAuthProviderFailure(undefined),
			});
			if (upstream.response.ok) return await providerJson(upstream);
			const status = upstream.response.status;
			await discardBody(upstream.response);
			throw new OAuthProviderFailure(status);
		} catch (error) {
			if (
				!(error instanceof OAuthProviderFailure) ||
				!retryableProviderFailure(error) ||
				attempt >= 2
			) {
				throw error;
			}
			await retryDelay(attempt);
		}
	}
	throw new OAuthProviderFailure(undefined);
}

async function providerFetch(
	url: string,
	options: ProviderRequestOptions,
): Promise<ProviderResponse> {
	const form = options.body instanceof URLSearchParams;
	const body =
		options.body instanceof URLSearchParams
			? options.body
			: JSON.stringify(options.body);
	const signal = providerSignal(options.signal);
	const networkError = options.networkError ?? (() => providerApiError());
	try {
		const response = await fetch(url, {
			method: "POST",
			headers: {
				Accept: "application/json",
				"Content-Type": form
					? "application/x-www-form-urlencoded"
					: "application/json",
			},
			body,
			signal,
		});
		return {
			response,
			...(options.signal ? { clientSignal: options.signal } : {}),
			networkError,
		};
	} catch (error) {
		if (options.signal?.aborted) throw clientAbortError(error);
		throw networkError();
	}
}

function providerSignal(clientSignal: AbortSignal | undefined): AbortSignal {
	const timeout = AbortSignal.timeout(PROVIDER_REQUEST_TIMEOUT_MS);
	return clientSignal ? AbortSignal.any([clientSignal, timeout]) : timeout;
}

async function requireProviderJson(upstream: ProviderResponse): Promise<unknown> {
	if (!upstream.response.ok) {
		await discardBody(upstream.response);
		throw providerApiError(upstream.response.status);
	}
	return providerJson(upstream);
}

async function providerJson(upstream: ProviderResponse): Promise<unknown> {
	let bytes: Uint8Array | null;
	try {
		bytes = await readLimitedBody(
			upstream.response,
			MAX_OAUTH_RESPONSE_BYTES,
		);
	} catch (error) {
		if (upstream.clientSignal?.aborted) throw clientAbortError(error);
		if (!(error instanceof BodySizeLimitError)) {
			throw upstream.networkError();
		}
		throw invalidProviderResponse();
	}
	if (!bytes) throw invalidProviderResponse();
	try {
		return JSON.parse(new TextDecoder().decode(bytes)) as unknown;
	} catch {
		throw invalidProviderResponse();
	}
}

function retryableProviderFailure(error: OAuthProviderFailure): boolean {
	return error.status === undefined || error.status === 429 || error.status >= 500;
}

function clientAbortError(error: unknown): DOMException {
	return isAbortError(error)
		? error
		: new DOMException("The client request was aborted.", "AbortError");
}

async function discardBody(response: Response): Promise<void> {
	try {
		await response.body?.cancel();
	} catch {
		// The response is already unusable; there is nothing sensitive to report.
	}
}

async function retryDelay(attempt: number): Promise<void> {
	await new Promise<void>((resolve) =>
		setTimeout(resolve, (attempt + 1) * 1000),
	);
}

function providerApiError(status?: number): ApiError {
	return new ApiError(
		502,
		"The OAuth provider request failed.",
		"upstream_error",
		status === 429 ? "oauth_rate_limited" : "oauth_provider_error",
	);
}

function invalidProviderResponse(): ApiError {
	return new ApiError(
		502,
		"The OAuth provider returned an invalid response.",
		"upstream_error",
		"invalid_oauth_provider_response",
	);
}

File: worker-rs/src/auth/oauth_provider.rs
//! OpenAI OAuth wire protocol and retry policy.

use serde_json::{Value, json};
use url::form_urlencoded;

use crate::core::{ApiError, AppResult};

use super::oauth_ports::{
    OAuthClock, OAuthHttpClient, OAuthHttpFailure, OAuthHttpMethod, OAuthHttpRequest,
    OAuthHttpResponse,
};

const CODEX_CLIENT_ID: &str = "app_EMoamEEZ73f0CkXaXp7hrann";
const DEVICE_USER_CODE_URL: &str = "https://auth.openai.com/api/accounts/deviceauth/usercode";
const DEVICE_TOKEN_URL: &str = "https://auth.openai.com/api/accounts/deviceauth/token";
pub const DEVICE_VERIFICATION_URL: &str = "https://auth.openai.com/codex/device";
const DEVICE_TOKEN_REDIRECT_URI: &str = "https://auth.openai.com/deviceauth/callback";
const OAUTH_TOKEN_URL: &str = "https://auth.openai.com/oauth/token";
pub const MAX_OAUTH_RESPONSE_BYTES: usize = 64 * 1024;
pub const PROVIDER_REQUEST_TIMEOUT_MS: u64 = 10_000;
const MAX_REFRESH_ATTEMPTS: usize = 3;

#[derive(Debug, Clone, PartialEq)]
pub struct ProviderDeviceAuthorization {
    pub device_auth_id: String,
    pub user_code: String,
    pub interval: Option<Value>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum ProviderDevicePollResult {
    Pending,
    Authorized { token_payload: Value },
}

pub struct OAuthProvider<'a> {
    http: &'a dyn OAuthHttpClient,
    clock: &'a dyn OAuthClock,
}

impl<'a> OAuthProvider<'a> {
    pub fn new(http: &'a dyn OAuthHttpClient, clock: &'a dyn OAuthClock) -> Self {
        Self { http, clock }
    }

    pub async fn request_device_authorization(&self) -> AppResult<ProviderDeviceAuthorization> {
        let response = self
            .execute_for_api(json_request(
                DEVICE_USER_CODE_URL,
                json!({"client_id": CODEX_CLIENT_ID}).to_string(),
            ))
            .await?;
        let payload = require_provider_json(response)?;
        let payload = payload.as_object().ok_or_else(invalid_provider_response)?;
        let device_auth_id =
            nonempty_string(payload.get("device_auth_id")).ok_or_else(invalid_provider_response)?;
        let user_code = nonempty_string(payload.get("user_code"))
            .or_else(|| nonempty_string(payload.get("usercode")))
            .ok_or_else(invalid_provider_response)?;
        Ok(ProviderDeviceAuthorization {
            device_auth_id: device_auth_id.to_owned(),
            user_code: user_code.to_owned(),
            interval: payload.get("interval").cloned(),
        })
    }

    pub async fn poll_device_authorization_token(
        &self,
        device_auth_id: &str,
        user_code: &str,
    ) -> AppResult<ProviderDevicePollResult> {
        let response = self
            .execute_for_api(json_request(
                DEVICE_TOKEN_URL,
                json!({
                    "device_auth_id": device_auth_id,
                    "user_code": user_code,
                })
                .to_string(),
            ))
            .await?;
        if matches!(response.status, 403 | 404) {
            return Ok(ProviderDevicePollResult::Pending);
        }
        let payload = require_provider_json(response)?;
        let payload = payload.as_object().ok_or_else(invalid_provider_response)?;
        let authorization_code = nonempty_string(payload.get("authorization_code"))
            .ok_or_else(invalid_provider_response)?;
        let code_verifier =
            nonempty_string(payload.get("code_verifier")).ok_or_else(invalid_provider_response)?;
        if nonempty_string(payload.get("code_challenge")).is_none() {
            return Err(invalid_provider_response());
        }
        let token_payload = self
            .exchange_authorization_code(authorization_code, code_verifier)
            .await?;
        Ok(ProviderDevicePollResult::Authorized { token_payload })
    }

    pub async fn refresh_provider_token(&self, refresh_token: &str) -> AppResult<Value> {
        for attempt in 0..MAX_REFRESH_ATTEMPTS {
            match self.exchange_refresh_token(refresh_token).await {
                Ok(payload) => return Ok(payload),
                Err(RefreshExchangeError::Api(error)) => return Err(error),
                Err(RefreshExchangeError::Provider(status))
                    if retryable_provider_failure(status) && attempt + 1 < MAX_REFRESH_ATTEMPTS =>
                {
                    self.clock.sleep_ms((attempt as u64 + 1) * 1_000).await;
                }
                Err(RefreshExchangeError::Provider(status)) => {
                    return Err(provider_api_error(status));
                }
            }
        }
        Err(provider_api_error(None))
    }

    async fn exchange_authorization_code(
        &self,
        code: &str,
        code_verifier: &str,
    ) -> AppResult<Value> {
        let response = self
            .execute_for_api(form_request(
                OAUTH_TOKEN_URL,
                &[
                    ("grant_type", "authorization_code"),
                    ("client_id", CODEX_CLIENT_ID),
                    ("code", code),
                    ("redirect_uri", DEVICE_TOKEN_REDIRECT_URI),
                    ("code_verifier", code_verifier),
                ],
            ))
            .await?;
        require_provider_json(response)
    }

    async fn exchange_refresh_token(
        &self,
        refresh_token: &str,
    ) -> Result<Value, RefreshExchangeError> {
        let request = form_request(
            OAUTH_TOKEN_URL,
            &[
                ("client_id", CODEX_CLIENT_ID),
                ("grant_type", "refresh_token"),
                ("refresh_token", refresh_token),
                ("scope", "openid profile email"),
            ],
        );
        let response = self.http.execute(request).await.map_err(|failure| {
            if failure == OAuthHttpFailure::ResponseTooLarge {
                RefreshExchangeError::Api(invalid_provider_response())
            } else {
                RefreshExchangeError::Provider(None)
            }
        })?;
        if !success_status(response.status) {
            return Err(RefreshExchangeError::Provider(Some(response.status)));
        }
        provider_json(&response.body).map_err(RefreshExchangeError::Api)
    }

    async fn execute_for_api(&self, request: OAuthHttpRequest) -> AppResult<OAuthHttpResponse> {
        self.http
            .execute(request)
            .await
            .map_err(|failure| match failure {
                OAuthHttpFailure::ClientAborted => request_aborted(),
                OAuthHttpFailure::ResponseTooLarge => invalid_provider_response(),
                OAuthHttpFailure::Network | OAuthHttpFailure::TimedOut => provider_api_error(None),
            })
    }
}

enum RefreshExchangeError {
    Api(ApiError),
    Provider(Option<u16>),
}

fn json_request(url: &str, body: String) -> OAuthHttpRequest {
    request(url, "application/json", body)
}

fn form_request(url: &str, values: &[(&str, &str)]) -> OAuthHttpRequest {
    let mut serializer = form_urlencoded::Serializer::new(String::new());
    serializer.extend_pairs(values.iter().copied());
    request(
        url,
        "application/x-www-form-urlencoded",
        serializer.finish(),
    )
}

fn request(url: &str, content_type: &str, body: String) -> OAuthHttpRequest {
    OAuthHttpRequest {
        method: OAuthHttpMethod::Post,
        url: url.to_owned(),
        headers: vec![
            ("Accept".into(), "application/json".into()),
            ("Content-Type".into(), content_type.into()),
        ],
        body,
        timeout_ms: PROVIDER_REQUEST_TIMEOUT_MS,
        max_response_bytes: MAX_OAUTH_RESPONSE_BYTES,
    }
}

fn require_provider_json(response: OAuthHttpResponse) -> AppResult<Value> {
    if !success_status(response.status) {
        return Err(provider_api_error(Some(response.status)));
    }
    provider_json(&response.body)
}

fn provider_json(body: &[u8]) -> AppResult<Value> {
    if body.is_empty() || body.len() > MAX_OAUTH_RESPONSE_BYTES {
        return Err(invalid_provider_response());
    }
    serde_json::from_slice(body).map_err(|_| invalid_provider_response())
}

fn success_status(status: u16) -> bool {
    (200..300).contains(&status)
}

fn retryable_provider_failure(status: Option<u16>) -> bool {
    status.is_none_or(|status| status == 429 || status >= 500)
}

fn nonempty_string(value: Option<&Value>) -> Option<&str> {
    value?.as_str().filter(|value| !value.is_empty())
}

fn request_aborted() -> ApiError {
    ApiError::new(408, "The request was cancelled or timed out.")
        .with_kind("request_timeout")
        .with_code("request_aborted")
}

fn provider_api_error(status: Option<u16>) -> ApiError {
    ApiError::new(502, "The OAuth provider request failed.")
        .with_kind("upstream_error")
        .with_code(if status == Some(429) {
            "oauth_rate_limited"
        } else {
            "oauth_provider_error"
        })
}

fn invalid_provider_response() -> ApiError {
    ApiError::new(502, "The OAuth provider returned an invalid response.")
        .with_kind("upstream_error")
        .with_code("invalid_oauth_provider_response")
}
